kmath-tensors/src/commonMain/kotlin/space/kscience/kmath/tensors/core/DoubleTensorAlgebra.kt

```diff
@@ -706,14 +705,377 @@ public open class DoubleTensorAlgebra :
+        }
+
+        val prnt          = opts[0]                // >1 intermediate results; >2 plots
+        val MaxIter       = opts[1].toInt()        // maximum number of iterations
+        val epsilon_1     = opts[2]                // convergence tolerance for gradient
+        val epsilon_2     = opts[3]                // convergence tolerance for parameters
+        val epsilon_3     = opts[4]                // convergence tolerance for Chi-square
+        val epsilon_4     = opts[5]                // determines acceptance of a L-M step
+        val lambda_0      = opts[6]                // initial value of damping paramter, lambda
+        val lambda_UP_fac = opts[7]                // factor for increasing lambda
+        val lambda_DN_fac = opts[8]                // factor for decreasing lambda
+        val Update_Type   = opts[9].toInt()        // 1: Levenberg-Marquardt lambda update
+                                                   // 2: Quadratic update
+                                                   // 3: Nielsen's lambda update equations
+
+        p_min = make_column(p_min)
+        p_max = make_column(p_max)
+
+        if (length(make_column(dp)) == 1) {
+            dp = ones(ShapeND(intArrayOf(Npar, 1))).div(1 / dp[0, 0]).as2D()
+        }
+
+        val idx = get_zero_indices(dp)                 // indices of the parameters to be fit
+        val Nfit = idx?.shape?.component1()            // number of parameters to fit
+        var stop = false                               // termination flag
+        val y_init = feval(func, t, p, settings)       // residual error using p_try
+
+        if (weight.shape.component1() == 1 || variance(weight) == 0.0) { // identical weights vector
+            weight = ones(ShapeND(intArrayOf(Npnt, 1))).div(1 / abs(weight[0, 0])).as2D()
+            // println("using uniform weights for error analysis")
+        }
+        else {
+            weight = make_column(weight)
+            weight.abs()
+        }
+
+        // initialize Jacobian with finite difference calculation
+        var lm_matx_ans = lm_matx(func, t, p_old, y_old,1, J, p, y_dat, weight, dp, settings)
+        var JtWJ = lm_matx_ans[0]
+        var JtWdy = lm_matx_ans[1]
+        X2 = lm_matx_ans[2][0, 0]
+        var y_hat = lm_matx_ans[3]
+        J = lm_matx_ans[4]
+
+        if ( abs(JtWdy).max()!! < epsilon_1 ) {
+//            println(" *** Your Initial Guess is Extremely Close to Optimal ***\n")
+//            println(" *** epsilon_1 = %e\n$epsilon_1")
+            stop = true
+        }
+
+        var lambda = 1.0
+        var nu = 1
+        when (Update_Type) {
+            1 -> lambda  = lambda_0                       // Marquardt: init'l lambda
+            else -> {                                     // Quadratic and Nielsen
+                lambda  = lambda_0 * (diag(JtWJ)).max()!!
+                nu = 2
+            }
+        }
```

Comment:
It could be an `if` if the algorithm has no other branches.

Reply:
fixed